### PR TITLE
[SPARK-24240] Add a config to control whether InMemoryFileIndex should update cache when refresh.

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?
In current code(https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala#L172), after data is inserted, spark will always refresh file index and update the cache, which costs extra overhead if the updated cache is not used any more.

This pr proposes to add a config -- `spark.sql.updateCacheWhenRefreshMemoryFileIndex`. When it's false, Spark only marks cache as outdated and do lazy cache-update when list the files.

## How was this patch tested?
Added a test.

### CODE REVIEW

**Comment by jinxing64** [comment]
cc @cloud-fan @gengliangwang @adrian-ionescu 
Do you have any thoughts? 
Please take a look when you have time, thanks !

**Comment by adrian-ionescu** [comment]
I just took a quick look... Do I understand correctly that your flag basically controls whether the refresh is eager (`true`, current behavior) vs lazy (`false`)? If so, let's rename the flag accordingly.

Also, with lazy mode we just delay file listing from immediately after an append operation until the next Spark job involving that table, right? It doesn't seem like we gain much overall..

What's the general use-case that motivated you to make this change? Please explain that more clearly in the PR description. Also, let's add a test that verifies the intended behavior.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 1]
```diff
@@ -75,16 +77,31 @@ class InMemoryFileIndex(
   }
 
   override protected def leafFiles: mutable.LinkedHashMap[Path, FileStatus] = {
+    refreshIfCacheOutdated()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: what's the gain here? seems you only delay the refresh, not eliminate it.

**Comment by jinxing64** [comment]
@cloud-fan @adrian-ionescu
Thanks for looking into this !
The issue is that when I append into a table with sql below:
```
insert into X select ....
```
It always refresh the file index of table X in memory, though I don't query X after this job. From this point, the refresh is not necessary and costs much time and memory.
In this pr, I proposes to mark the file index as 'outdated', and postpone the refresh until next query of the file index.

**Comment by jinxing64** [comment]
I mean the memory refresh always happen after writing data, which I think is not necessary ?

**Comment by adrian-ionescu** [comment]
Fair enough. Actually I think that's how `CatalogFileIndex` works. So, for consistency, we might even want to consider making this the default.

Please add a test, though, and improve the PR description.

**Comment by jinxing64** [comment]
@cloud-fan  @adrian-ionescu 
I added a test, please check when you have time.

**Pull request closed without merging**